File: client/preview.lua
local QBCore = exports['qb-core']:GetCoreObject()

RegisterNetEvent("tr_patrolvehicles:preview")
AddEventHandler("tr_patrolvehicles:preview", function(data)
    QBCore.Functions.TriggerCallback('tr_patrolvehicles:CheckIfActive', function(status)
        if status then
            local preview = data.preview
            local cam = data.preview.cam
            local coords = preview.coords
            InPreview = true
            QBCore.Functions.SpawnVehicle(data.vehicle, function(veh)
                FreezeEntityPosition(PlayerPedId(), true)
                SetVehicleNumberPlateText(veh, "PREVIEW")
                exports[Config.FuelSystem]:SetFuel(veh, 0.0)
                FreezeEntityPosition(veh, true)
                SetVehicleEngineOn(veh, false, false, true)
                DoScreenFadeOut(200)
                Wait(500)
                DoScreenFadeIn(200)
                SetVehicleUndriveable(veh, true)

                VehicleCam = CreateCamWithParams("DEFAULT_SCRIPTED_CAMERA", cam.coords.x, cam.coords.y, cam.coords.z, cam.rotation.verticalrotate, cam.rotation.horizontalrotate, cam.rotation.left_n_right, cam.fov, false, 0)   
                SetCamActive(VehicleCam, true)
                RenderScriptCams(true, true, 2000, true, true)
                
                CreateThread(function()
                    while true do
                        if InPreview then
                            exports['qb-core']:DrawText('⇽', 'bottom')
                        elseif not InPreview then
                            exports['qb-core']:HideText()
                            break
                        end
                        Wait(1)
                    end
                end)
        
                CreateThread(function()
                    while true do
                        if IsControlJustReleased(0, 177) then
                            FreezeEntityPosition(PlayerPedId(), false)
                            QBCore.Functions.DeleteVehicle(veh)
                            DoScreenFadeOut(200)
                            Wait(500)
                            DoScreenFadeIn(200)
                            RenderScriptCams(false, false, 1, true, true)
                            InPreview = false
                            TriggerServerEvent("tr_patrolvehicles:server:SetActive", false)
                            break
                        end
                        Wait(1)
                    end
                end)
            end, coords, true)
        end
    end)
end)

File: client/modules/preview.js
let isInPreview = false

function createPreviewCam(key, netId) {
    const vehicle = NetworkGetEntityFromNetworkId(netId)
    const cam = System[key].VEHICLES.preview.cam
    Bridge.drawText.show()
    isInPreview = true

    FreezeEntityPosition(PlayerPedId(), true)
    SetVehicleUndriveable(vehicle, true)

    DoScreenFadeOut(200)
    setTimeout(() => {
        DoScreenFadeIn(200)
    }, 500)
    camHandle = CreateCamWithParams("DEFAULT_SCRIPTED_CAMERA", cam.coords[0], cam.coords[1], cam.coords[2], cam.rotation.verticalrotate, cam.rotation.horizontalrotate, cam.rotation.left_n_right, cam.fov, false, 0)
    SetCamActive(camHandle, true)
    RenderScriptCams(true, true, 2000, true, true)
}

async function clearPreviewCam(netId) {
    Bridge.drawText.hide()

    FreezeEntityPosition(PlayerPedId(), false)
    const success = await exports.tr_kit.clearCreatedVehicle(netId)
    if (!success) lib.console.err('failed to create the preview vehicle, reponse was: ' + success)

    DoScreenFadeOut(200)
    setTimeout(() => {
        DoScreenFadeIn(200)
    }, 500)
    RenderScriptCams(false, false, 2000, true, true)
    isInPreview = false
}

async function PreviewVehicle(key, index) {
    const selectedConfig = System[key].ITEM
    const configItems = Items[selectedConfig]
    if (!isInPreview) {
        const isPreviewSessionClear = await lib.callback.await('sessionStatus')
        if (!isPreviewSessionClear) {
            Bridge.notify('Preview service is currently busy, try again later')
            return
        }
        const isThePlaceClear = await IsZoneFree(System[key].VEHICLES.preview.coords)
        if (!isThePlaceClear) {
            Bridge.notify('Preview point is currently busy!', 'error')
            return
        }
        emitNet('lenix_vehicles:server:setPreviewSessionBusy', true)
        const netId = await exports.tr_kit.createSingleVehicle({
            hash: GetHashKey(configItems[index].vehicle),
            coords: System[key].VEHICLES.preview.coords,
        })
        if (netId) {
            createPreviewCam(key, netId)
            const tick = setTick(() => {
                if (IsControlJustReleased(0, 177)) {
                    emitNet('lenix_vehicles:server:setPreviewSessionBusy', false)
                    clearPreviewCam(netId);
                    clearTick(tick);
                }
            });
        }
    }
}